src/box/box.cc

```diff
@@ -1293,6 +1310,39 @@ box_check_replicaset_uuid(struct tt_uuid *uuid)
 	return box_check_uuid(uuid, "replicaset_uuid");
 }
 
+/** Check bootstrap_leader option validity. */
+static int
+box_check_bootstrap_leader(struct uri *uri, struct tt_uuid *uuid)
+{
+	*uuid = uuid_nil;
+	uri_create(uri, NULL);
+	const char *source = cfg_gets("bootstrap_leader");
+	enum bootstrap_strategy strategy = box_check_bootstrap_strategy();
+	if (strategy != BOOTSTRAP_STRATEGY_CONFIG) {
+		if (source == NULL) {
+			/* Nothing to do. */
+			return 0;
+		}
+		diag_set(ClientError, ER_CFG, "bootstrap_leader",
+			 "the option takes no effect when bootstrap strategy "
+			 "is not 'config'");
+		return -1;
+	} else if (source == NULL) {
+		diag_set(ClientError, ER_CFG, "bootstrap_leader",
+			 "the option can't be empty when bootstrap strategy "
+			 "is 'config'");
+		return -1;
+	}
+	if (box_check_uri(uri, "bootstrap_leader") == 0)
+		return 0;
+	/* Not a uri. Try uuid then. */
```

Comment:
The error wasn't critical, but the counter in `box.stat` is incremented. Maybe better find a way not to set a diag.

Reply:
Yes, I missed that, thanks!
What do you think of this approach?
<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index f64b7c2ab..4e1c3932a 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -984,7 +984,7 @@ box_check_election_fencing_mode(void)
 
 /** A helper to check validity of a single uri. */
 static int
-check_uri(const struct uri *uri, const char *option_name)
+check_uri(const struct uri *uri, const char *option_name, bool set_diag)
 {
 	const char *auth_type = uri_param(uri, "auth_type", 0);
 	const char *errmsg;
@@ -999,28 +999,32 @@ check_uri(const struct uri *uri, const char *option_name)
 	}
 	return 0;
 bad_uri:;
-	char *uristr = tt_static_buf();
-	uri_format(uristr, TT_STATIC_BUF_LEN, uri, false);
-	diag_set(ClientError, ER_CFG, option_name,
-		 tt_sprintf("bad URI '%s': %s", uristr, errmsg));
+	if (set_diag) {
+		char *uristr = tt_static_buf();
+		uri_format(uristr, TT_STATIC_BUF_LEN, uri, false);
+		diag_set(ClientError, ER_CFG, option_name,
+			 tt_sprintf("bad URI '%s': %s", uristr, errmsg));
+	}
 	return -1;
 }
 
 /** Check validity of a uri passed in configuration option. */
 static int
-box_check_uri(struct uri *uri, const char *option_name)
+box_check_uri(struct uri *uri, const char *option_name, bool set_diag)
 {
 	const char *source = cfg_gets(option_name);
 	if (source == NULL) {
 		uri_create(uri, NULL);
 		return 0;
 	}
-	if (uri_create(uri, source) != 0) {
+	if (uri_create(uri, source) == 0)
+		return check_uri(uri, option_name, set_diag);
+	if (set_diag) {
 		diag_set(ClientError, ER_CFG, option_name,
 			 tt_sprintf("bad URI '%s': expected host:service or "
 				    "/unix.socket", source));
 	}
-	return check_uri(uri, option_name);
+	return -1;
 }
 
 static int
@@ -1035,7 +1039,7 @@ box_check_uri_set(const char *option_name)
 	int rc = 0;
 	for (int i = 0; i < uri_set.uri_count; i++) {
 		const struct uri *uri = &uri_set.uris[i];
-		if (check_uri(uri, option_name) != 0) {
+		if (check_uri(uri, option_name, true) != 0) {
 			rc = -1;
 			break;
 		}
@@ -1256,19 +1260,22 @@ box_check_replication_sync_timeout(void)
 }
 
 static inline int
-box_check_uuid(struct tt_uuid *uuid, const char *name)
+box_check_uuid(struct tt_uuid *uuid, const char *name, bool set_diag)
 {
 	*uuid = uuid_nil;
 	const char *uuid_str = cfg_gets(name);
 	if (uuid_str == NULL)
 		return 0;
 	if (tt_uuid_from_string(uuid_str, uuid) != 0) {
-		diag_set(ClientError, ER_CFG, name, uuid_str);
+		if (set_diag)
+			diag_set(ClientError, ER_CFG, name, uuid_str);
 		return -1;
 	}
 	if (tt_uuid_is_nil(uuid)) {
-		diag_set(ClientError, ER_CFG, name,
-			 tt_sprintf("nil UUID is reserved"));
+		if (set_diag) {
+			diag_set(ClientError, ER_CFG, name,
+				 tt_sprintf("nil UUID is reserved"));
+		}
 		return -1;
 	}
 	return 0;
@@ -1303,13 +1310,13 @@ box_check_replication_anon(void)
 static int
 box_check_instance_uuid(struct tt_uuid *uuid)
 {
-	return box_check_uuid(uuid, "instance_uuid");
+	return box_check_uuid(uuid, "instance_uuid", true);
 }
 
 static int
 box_check_replicaset_uuid(struct tt_uuid *uuid)
 {
-	return box_check_uuid(uuid, "replicaset_uuid");
+	return box_check_uuid(uuid, "replicaset_uuid", true);
 }
 
 /** Check bootstrap_leader option validity. */
@@ -1335,10 +1342,10 @@ box_check_bootstrap_leader(struct uri *uri, struct tt_uuid *uuid)
 			 "is 'config'");
 		return -1;
 	}
-	if (box_check_uri(uri, "bootstrap_leader") == 0)
+	if (box_check_uri(uri, "bootstrap_leader", false) == 0)
 		return 0;
 	/* Not a uri. Try uuid then. */
-	if (box_check_uuid(uuid, "bootstrap_leader") == 0)
+	if (box_check_uuid(uuid, "bootstrap_leader", false) == 0)
 		return 0;
 	diag_set(ClientError, ER_CFG, "bootstrap_leader",
 		 "the value must be either a uri or a uuid");
```
</details>

Reply:
Looks ugly. But I can't see a better option.

Reply:
Another variant would be to make `box_check_uri` and `box_check_uuid` to return error strings like this:
`box_check_uuid(struct tt_uuid *uuid, const char *name, char **errmsg)` and not set diag at all.
The caller of `box_check_uuid/uri` can then decide whether to use the error message anyhow (like set diag) or ignore it.
I'm not sure it's any better than the variant above though.